### ISSUE TITLE
AudioUnitRender error on macOS

### ISSUE THREAD

**Issue by monaghanwashere** [kcat/openal-soft#750]
I'm hitting a bug where, every so often, OpenAL Soft fails to capture input samples, and keeps spamming the following to the console:

`[ALSOFT] (EE) AudioUnitRender error: -10877`

I did a bit of research and it seems like this is `kAudioUnitErr_InvalidElement`, which is returned when an AudioUnitRender call passes an invalid index for the bus, which is argument # 4 here:

https://github.com/kcat/openal-soft/blob/f2858ac8656c22e8d300cba42883bf3008206c1c/alc/backends/coreaudio.cpp#L598

It seems like `audiobuf.list.mNumberBuffers` is `1` or `2` depending on the amount of data in the input buffer, but I'm not sure exactly how this is determined.

I'm opening the capture device like so:
```
const ALCchar* defaultDeviceName = alcGetString(NULL, ALC_CAPTURE_DEFAULT_DEVICE_SPECIFIER);
ALCdevice *mDevice = alcCaptureOpenDevice(defaultDeviceName, 48000, AL_FORMAT_MONO_FLOAT32, 12000);
if (!mDevice) return;
alcCaptureStart(mDevice);
ALCint samplesIn = 0;
```
And this callback gets invoked every so often to obtain the captured samples:
```
void getCaptureBuffer(AudioCaptureBuffer& data)
{
    ALCint samplesIn = 0;
    alcGetIntegerv(mDevice, ALC_CAPTURE_SAMPLES, 1, &samplesIn);
    int frameSize = 1024;
 
    while (samplesIn > frameSize)
    {
        // Grab the sound
        alcCaptureSamples(mDevice, mFrameBuffer.data(), frameSize);

        // Put into ring buffer
        for (int sample = 0; sample < frameSize; sample++)
        {
            mBuffer.write(mFrameBuffer[sample]);
        }

        alcGetIntegerv(mDevice, ALC_CAPTURE_SAMPLES, 1, &samplesIn);
    }

    data = mBuffer;
}
```

My capture code is presumably fine, because a) I only hit this error occasionally (seemingly randomly), and b) I don't have this issue on Windows (where the capture code is identical).

Any ideas?

**Comment by kcat** [comment]
It seems I mishandled the parameter for `AudioUnitRender`. I'm not sure why I thought to pass `audiobuf.list.mNumberBuffers` (which is `1` if the number of samples being captured can fit on the first section of the ring buffer, or `2` if it needs to capture more than that and needs to wrap around) for the bus number.

Try now with commit 832ccd7f489f7db763a164c663d730f4cadb71bc. It passes along the bus number given to the callback to the `AudioUnitRender` call, which should hopefully be correct. A closer look also shows the callback gets a `AudioUnitRenderActionFlags*`, and the `AudioUnitRender` call takes one. I'm not sure if that should be passed along too, or what misbehavior may occur if it's given the wrong thing.

**Comment by monaghanwashere** [comment]
Commit now gives the following error instead:

`[ALSOFT] (EE) AudioUnitRender error: -50`

Which seems to correspond to `kAudio_ParamError`.

**Comment by kcat** [comment]
Does it always give that error, or only sometimes like when it was returning `kAudioUnitErr_InvalidElement`?

**Comment by monaghanwashere** [comment]
It always gives the error i.e. every time I make the `alcCaptureSamples` call to capture samples.

**Comment by kcat** [comment]
Try now with commit 4115df49cfe1e3d0bc1723996c90d0a5bcf1f1f3.

**Comment by monaghanwashere** [comment]
No luck, unfortunately. Still the same error (`-50`).

To try and see if there is a difference between `alcCaptureStart` and `alcCaptureSamples`, I commented out my `alcCaptureSamples` call (which I perform every frame), and I'm noticing the following:
- After calling `alcCaptureStart`, AudioUnitRender error `-10863` is continuously happening. This seems to be `kAudioUnitErr_CannotDoInCurrentContext`.
- Putting back the subsequent `alcCaptureSamples` call is what then causes AudioUnitRender error `-50` to happen every frame.

**Comment by kcat** [comment]
> After calling `alcCaptureStart`, AudioUnitRender error `-10863` is continuously happening. This seems to be `kAudioUnitErr_CannotDoInCurrentContext`.

Hmm, perhaps it doesn't like being asked to "render"/capture 0 samples once OpenAL Soft's ringbuffer has no more room.

> Putting back the subsequent `alcCaptureSamples` call is what then causes AudioUnitRender error `-50` to happen every frame.

Not sure about why this happens. The call to capture samples would mean the ringbuffer is read from and more samples can keep getting captured into it. Perhaps `AudioUnitRender` doesn't like being given 2 buffers when OpenAL Soft's ringbuffer wraps around. If that's the case, I'm not sure how to make CoreAudio capture the samples into two separate buffer segments instead of one.

**Comment by kcat** [comment]
Does it work now with commit 19d35a45d7f1a7348650b824b98a11dd6c51e49f?

**Comment by monaghanwashere** [comment]
Interesting, now with commit ['19d35a4'](https://github.com/kcat/openal-soft/commit/19d35a45d7f1a7348650b824b98a11dd6c51e49f), error `-10863` seems to have disappeared completely, but error `-50` persists. I'm seeing `AudioUnitRender capture error: -50` happen continuously after a (single) call to `alcCaptureStart`.

**Comment by kcat** [comment]
I'm completely baffled by that. Nothing about the `AudioUnitRender` call is changed when some samples are read, it just changes if or how many samples can be written into the ringbuffer afterward.

I did notice the write amount for the ringbuffer was wrong, and is fixed now in commit 07928fbb6436b24b28e49e0590e62588319fe78e, but that shouldn't change the behavior of that call.

**Comment by monaghanwashere** [comment]
Just for good measure, I tested [`07928fb`](https://github.com/kcat/openal-soft/commit/07928fbb6436b24b28e49e0590e62588319fe78e) as well, and yeah, same error unfortunately.

**Comment by monaghanwashere** [comment]
I've discovered some extra information -- there is another library running in the application along side openal soft, namely webrtc, which also seems to be doing some audio capture/processing. I've sifted through the relevant webrtc source code on macOS, and it seems it too requests and works with an AudioUnit on the system (see [here](https://webrtc.googlesource.com/src/+/refs/heads/main/sdk/objc/native/src/audio/voice_processing_audio_unit.mm#89)). I tried running the application with webrtc disabled, and in that scenario, openal soft does **_not_** produce any errors and I am able to successfully capture audio samples.

I don't see why openal soft should misbehave when webrtc is enabled, however. Does this information help provide you with any clues about what's going on with openal soft?

**Comment by kcat** [comment]
> I don't see why openal soft should misbehave when webrtc is enabled, however. Does this information help provide you with any clues about what's going on with openal soft?

Do you have multiple recording devices on the system? If so, do you still get errors if webrtc and OpenAL Soft are capturing from different devices? Perhaps macOS doesn't like a single process having multiple active recording units for the same device.

**Comment by monaghanwashere** [comment]
Yes, I have multiple recording devices on the system. I tried having OpenAL Soft and webrtc capture from different devices as you suggested, and I indeed do _*not*_ see the OpenAL Soft errors any more. How can I confirm that this is a macOS property? I don't have any errors ordinarily in other situations (egs. using both Google Hangouts, and Zoom, at the same time, with both recording concurrently)

**Comment by kcat** [comment]
Do you still get errors if OpenAL Soft and webrtc are capturing from the same device in different processes? i.e. one app capturing using webrtc, and another app capturing from the same device using OpenAL Soft (without webrtc). If there are errors, it would seem to be a limitation of CoreAudio that it can't resolve the differences in how webrtc and OpenAL Soft handle the device (or I'm doing something wrong in OpenAL Soft, though I wouldn't know where to begin looking). If there are no errors, it would seem to be a bug or shortcoming of CoreAudio that multiple uses of the same device in one process results in errors.
